course/receivers.py

```diff
@@ -109,4 +109,40 @@ def may_preapprove_role(course, user):
 
 # }}}
 
+
+@receiver(post_delete, sender=FlowSession)
+@transaction.atomic
+def create_exempt_grade_change_when_delete_session(sender, instance, **kwargs):
+    # type: (Any, FlowSession, **Any) -> None
+    """
+    Create a :class:`GradeChange` entry with state "exempt" for the attempt_id
+    of that flow_session, after deleting a flow session from admin. We are
+    strongly against modifying data through admin, but this is helpful for
+    debugging when developing.
+    """
+    from course.flow import get_flow_session_attempt_id
+    last_gchanges = (
+        GradeChange.objects
+        # query using flow_session=instance will fail.
+        .filter(attempt_id=get_flow_session_attempt_id(instance))
+        .order_by("-grade_time")[:1])
+
+    if not last_gchanges.count():
+        return
+
+    last_gchange, = last_gchanges
+
+    from course.constants import grade_state_change_types
+
+    last_gchange.pk = None
+    last_gchange.points = None
+    last_gchange.flow_session = None
+    last_gchange.creator = None
+    last_gchange.comment = None
+
+    from django.utils.timezone import now
+    last_gchange.grade_time = now()
+    last_gchange.state = grade_state_change_types.exempt
+    last_gchange.save()
```

Comment:
"Exempt" is intended to mean that a student does not need to complete whatever is required for that grading opportunity, or at least that's the intention. (Yeah, I know--not well-documented.) Why is that the correct thing to do here?

Reply:
(as opposed to a do-over, for example)

Reply:
> "Exempt" is intended to mean that a student does not need to complete whatever is required for that grading opportunity,

Sorry, I misunderstand that.

My idea is: since the session was deleted,
1. existing grade changes of that session should not be counted into the overall grade, thus we need to clear all grade changes related.
1. there should be no more NEW related grade changes for that session (or we throw an error),

Reply:
Inconsistency were observed for a610287d30a85e79be051bbc757099579cd260e6.

Reply:
For this intentsion:

> "Exempt" is intended to mean that a student does not need to complete whatever is required for that grading opportunity, or at least that's the intention. 

P.S., I felt the currently it's not working this way, when new grades are to the opportunity, it throw [errors](https://github.com/inducer/relate/blob/816f832f87f2568df422593d3fd8a8541ac7f795/course/models.py#L1653-L1656).

Reply:
> currently, when new grades are append to the opportunity which already had a exempt gchange , it throw errors.

No, that's expected. "Exempt" is intended to mean, e.g. a student was sick and does no longer need to complete the assignment. As such, it is not expected that they complete the assignment, and it is considered a problem if they do. (because then the "exemption" was meaningless) Another plausible behavior would be to ignore any grade changes after "exempt" hits.